test/unit/range/iterator_test_template.hpp

```diff
@@ -0,0 +1,502 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <seqan3/core/type_traits/range.hpp>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+
+using namespace seqan3;
+
+template <typename T>
+struct iterator_fixture : public ::testing::Test
```

Comment:
can this class not simply inherit from T and just forward declare the members?
Then you can do the sanity check inside of the fixture, instead of providing a test for it, which actually tests nothing but either compiles or not. The user then just provides it's type instead of specializing the iterator_fixture.

```cpp
template <typename fixture_t>
struct iterator_fixture : public ::testing::Test, public fixture_t
{
    using iterator_tag = typename fixture_t::iterator_tag;
    using fixture_t::range_to_comapre;
    using fixture_t::begin_iterator;
    ...
    
    template <typename A, typename B>
        requires requires (A a, B b){ fixture_t::expect_eq(a, b) -> bool } && !EqualityComparableWith<A, B>
    static bool expect_eq(A && a, B && b)
    { EXPECT_TRUE((fixture_t::expect_eq(a, b))); }

    template <typename A, typename B>
        requires EqualityComparableWith<A, B>
    static bool expect_eq(A && a, B && b)
    { EXPECT_EQ(a, b); }
};
```

I never tested multiple inheritance in the google test framework but I don't know why that should not work. I find this more natural and clearer for the developer to use. Like test_equal is now part of the fixture and not a free function, that might be defined somewhere else, etc.

Reply:
Also basically everything that was a comment before can now be actual code.

Reply:
:worried: I still disagree

In your solution how do I handle the begin_const_iterator that only needs to be present if the variable const_iterable is set to true. I cannot constrain member variables? I would therefore need getter functions that are constrained which would blow up the code (especially if I would then do this constistently also with the normal begin_iterator). I would also need to write a concept `iterator_fixture_concept` that constrains the fixture_t, if I don't want to do that in the sanity_chack_test. And as this involes data structures and inheritance, following the not-discussed-detail-test-approach I would need to test that the fixture inherits corrrectly :laughing:. 
So I need another 2-3 hour to make all the changes although the current solution also works and has all neccessary constraints?

(In addition, for me this is less intuitive because I am specialising an actual fixture with made up a fixture. Why the extra level of abstraction when I can directly specify the fixture by reading the comments.)

Reply:
Ads discussed offline. We will keep it for now, although I am not a 100% fan of the current design, though in the end it is just a test and I don't care to much.